

# Introduction: Chip-Chip

```{r options, echo=FALSE}
library(knitr)

```

```{r setup,echo=FALSE,results="hide",message=FALSE}
suppressPackageStartupMessages({library(GenomicFeatures)
library(GenomicRanges)
library(IRanges)})
```


## Interval data on Estrogen Receptor binding

The Rmd for this file can be found
[here](https://github.com/gurinina/omic_sciences/blob/main/12-Chip-chip.Rmd).

A 2006 Nature Genetics paper by Carroll, Meyer, Song et al. describes
the notion that a group of genes whose expression correlates with the
expression of the estrogen receptor (ER) gene forms a signature of a
breast cancer subtype called "luminal". This finding motivates interest
in determining the set of all genomic elements to which ER binds, and
this has been carried out using ChIP-seq experiments. An important
finding of the Carroll et al. paper was that ER binding in the MCF-7
cell line occurred only rarely at promoter-proximal regions. This study
therefore unearthed previously unexplored cis-regulatory sites.

We will show how to relate binding peak locations to gene transcription
start sites in two cell lines analyzed in the ENCODE project.

**Acquisition of binding data**

The examples shown here are reported binding sites for estrogen related
receptor binding sites derinved from ChIPSeq experiments on two of the
main cell lines used by ENCODE: HepG2 and GM12878. In this section we
focus on the analysis that comes after the genomic regions are defined.
These data can be found as NarrowPeak files on the public repositories
but we have created a R package with `GRanges` objects already created.
If you have not done so already you can install it like this:

```{r geter,eval=FALSE}
suppressPackageStartupMessages({library(devtools)})
install_github("genomicsclass/ERBS")
```

Once installed you can attach two object, one for each cell line, and
view it's content:

```{r}
suppressPackageStartupMessages({library(ERBS)})
data(HepG2)  # cell line of liver origin
data(GM12878)  # immortalized B cell
HepG2  # locations of ER binding peaks
```

Note that these object contain two main parts, the regions of the genome
which we can extract with `granges`:

```{r}
granges(HepG2)
```

and other information that we can obtain with either the `mcols` or
`values` functions:

```{r}
values(HepG2) ##or mcols(HepG2)
```

<a name="basic"></a> \## Basic operations

These object are of class `GRanges` defined in the `GenomicsRanges`
package:

```{r}
class(HepG2)
```

Here we start with some very basic example. Suppose these regions are
ordered and we want to work with just the first 10. We can subset these
objects as you would expect:

```{r}
granges( HepG2[1:10] )
```

Now what if we want to subset the object by chromosomes? We can access
the chromosomes with `seqnames`. Note that not all organisms have
chromosomes so Bioconductor uses a more genreal name:

```{r}
chr <- seqnames(HepG2)
chr
```

Note that chr is not a factor or character vector as we would expect.
Instead run-length encoding is used. The reason for this is that it
provides a much more efficient way to store large granges, for example
those obtained from stroing short reads. Note also that we have
chromosome names that are not the standard chromosome names such as
`chrUn_gl000247`. The human genome actually has some sequences that have
not been mapped into one of the chromosomes and are given names like
these.

The Rle class behaves like factors in several useful ways. For example
we can tabulate:

```{r}
table(chr)[1:24]
```

We can also subset the object to data based on these. Here is the subset
of regions on chr20:

```{r}
granges( HepG2[ chr == "chr20" ] )
```

## Genomic Ranges tools

Before we continue on to finding gene near our binding sites we are
going to construct a consensus `GRanges`. Specifically, we are going to
construct an object representing the regions that are reported for both
cell lines. We can easily find these regions using the `findOverlaps`
function:

```{r}
res = findOverlaps(HepG2,GM12878)
res
```

```{r}
erbs = HepG2[queryHits(res)]
erbs = granges(erbs)
```

## Acquisition of gene transcription start sites

We can use Homo.sapiens:

```{r geths}
library(Homo.sapiens)
ghs = genes(Homo.sapiens)
ghs
```

Find nearest genes -- e.g. map the genes in erbs

```{r}
index <- GenomicRanges::precede(erbs,ghs)
erbs[1:3,]
granges(erbs)[1:3,]
granges(ghs)[index[1:3],]
d = distance(erbs,ghs[index])
```

If we want the closest to a start site, regardless of before or after,
for example if overlaps. We want something different.

We can create a granges of TSS like this:

Explain genes have direction and what is a transcription start site

```{r}
tssgr = resize(ghs,1)
granges(ghs)[1:3,]
granges(tssgr)[1:3,]
```

## Finding nearest gene for each binding event

The distanceToNearest function from GenomicFeatures will create a Hits
object with metadata giving the positive distance between peak location
and nearest TSS. We'll discuss the Hits object in the video.

```{r donear}
d = distanceToNearest(erbs, tssgr)
d
##note this gives an error
try(d[,3])
###instead we have to do this
dists = values(d)$distance
```

We will call a distance negative if the peak is 5' to the nearest TSS.
The density estimate given below shows that the vast majority of events
are remote from the 1kb region around TSS; we use dashed lines to denote
that region.

```{r getsi,fig=TRUE}
index = subjectHits(d)
sdists = ifelse(end(erbs) < start(tssgr[index]), dists, -dists)
hist(sdists, xlim=c(-100000,100000), main="Density of d(ER binding peak, nearest TSS)" ,breaks=seq(min(sdists),max(sdists),len=1000))
abline(v=-c(10000,10000), lty=2)
```

Getting information about genes

**Annotating genes**

Get information on genes

```{r}
index <- subjectHits(d)[dists<1000]
dists = values(d)$distance
##better way to do this?
geneids <- mcols(tssgr[index])$GENEID
```

```{r geths2}
library(Homo.sapiens)
??select
columns(Homo.sapiens)
keytypes(Homo.sapiens)
geneids <- as.character(geneids)

columns(Homo.sapiens)
df = select(Homo.sapiens,keys=geneids,columns=c("GENENAME","SYMBOL"),keytype="GENEID")
df
```

So we've now mapped the DNA binding sites from Chip-chip to the nearest
transcription start sites, and returned information about these genes,
including the gene name and gene symbol using the gene id.

88888888888888888888888888888888888888888888888888888888888888

**More formal introduction to ChIP-seq**

ChIP-seq is a protocol for inferring the locations of proteins bound or
associated with DNA. The raw data looks quite different than DNA- or
RNA-seq, in that the NGS reads form tall "peaks" at the locations where
the proteins were tightly bound to DNA in the cells which were used to
create the sample. More specifically, ChIP-seq results in two peaks of
reads of different strands (plus/minus also referred to as
Watson/Crick), as shown in [Figure
1](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2592715/figure/F1/) of
the MACS manuscript: [Zhang 2008](#foot)

## Peak calling

There are many different algorithms for calling peaks, which have
varying performance on different kinds of experiments. As mentioned in
the lecture, for ChIP of proteins with broad peaks (such as modified
histones), algorithms other than those for detecting sharp peaks might
perform better.

**After peak calling**

A number of analyses might be of interest following peak calling. In
this lab we will focus on differential binding across samples, by
focusing on the peak regions and counting the number of ChIP-seq reads
which fall into the peaks for each sample.

Motif-finding is common ChIP-seq analysis which is not explored in this
course, as we do not cover the basics of analysis of sequences.
Motif-finding refers to the task of looking for common strings of DNA
letters contained within peaks. These are biologically meaningful, as a
number of proteins which are bound to DNA have conformations which make
certain strings of DNA letters more preferable for binding. For more
references, see the [Footnotes](#foot).

## Differential binding across samples

The following lab will go over the functionality of the `DiffBind`
package, mostly using code from the vignette. This package is useful for
manipulating ChIP-seq signal in R, for comparing signal across files and
for performing tests of diffential binding.

**Reading peak files into R**

We check the files in the `DiffBind` folder, and in the `peaks`
subdirectory:

```{r}

dirwd = getwd()
library(DiffBind)
setwd(system.file("extra", package="DiffBind"))
list.files()
read.csv("tamoxifen.csv")
list.files("peaks")
```

The `dba` function creates the basic object for an analysis of
*Differential Binding Affinity*. The sample sheet specifies a data frame
of file with certain required columns. Note that columns have restricted
names, including *Tissue*, *Factor*, *Condition*, etc., which will be
referred to later in analysis.

This function will automatically create a correlation plot showing the
overlap of the peaks for all the samples.

```{r}

setwd(system.file("extra", package="DiffBind"))
ta <- dba(sampleSheet="tamoxifen.csv")
ta
```

From the `DiffBind` vignette, we have:

> This shows how many peaks are in each peakset, as well as (in the
> first line) total number of unique peaks *after merging overlapping
> ones* (3557) and the default binding matrix of 11 samples by the 2602
> sites that *overlap in at least two of the samples*."

We can access the peaks for each file:

```{r}

names(ta)
class(ta$peaks)
head(ta$peaks[[1]])
```

**Differential binding**

The following code chunk will count the reads from the BAM files
specified in the `samples` slot:
```{r,fig.height=300,fig.width=300,fig.cap ="correlation of samples"}

par(mfrow=c(1,1))
par(mar=c(1,1,1,1))
ta$samples
# this call does not actually work, because the BAM files are not included in the package
#ta <- dba.count(ta, minOverlap=3)
# instead we load the counts:
data(tamoxifen_counts)
ta2 <- tamoxifen
correlations <- DiffBind:::pv.occupancy(ta2,
    Sort = "cor", bCorOnly = TRUE, CorMethod = "pearson")
d=as.data.frame(correlations)
mm=matrix(rep(NA,121),nrow=11,ncol=11)
for(i in 1:55) mm[d$A[i],d$B[i]]=d$Cor[i]
mm[lower.tri(mm)]=t(mm)[lower.tri(t(mm))]
diag(mm)=1
rownames(mm)=ta2$samples$SampleID
colnames(mm)=ta2$samples$SampleID

nn = ta2$samples[,c("Tissue","Replicate")]
library(RColorBrewer)
library(rafalib)
reds = brewer.pal(4,"Reds")
purps =  brewer.pal(3,"Purples")
cols <- colorRampPalette(brewer.pal(9, "Greens"))(256)
nn$Tissue =reds[as.fumeric(nn$Tissue)]
nn$Replicate = purps[nn$Replicate]
DiffBind:::heatmap.3(mm, dendrogram = "column", symm = TRUE, scale = "none",
key=FALSE, col = cols,trace =  "none", margins = c(6,6), ColSideColors=as.matrix(nn),Colv=TRUE) 
#dba.plotHeatmap(ta2,margin=c(10),dendrogram ="column",key=FALSE) 
```

We can perform a test by specifying to contrast over the levels of
condition. This will call edgeR (the default) or DESeq software in order
to normalize samples for sequencing depth and perform essentially the
same analysis as a differential expression analysis for RNA-Seq counts.
Here we could also include the tissue as a blocking factor, by providing
`DBA_TISSUE` to the `block` argument of `dba.contrast`.

The plot produced then looks at correlation only for those peaks which
showed evidence of differential binding.

```{r}
ta2 <- dba.contrast(ta2, categories=DBA_CONDITION)
ta2 <- dba.analyze(ta2)
ta2
```

From the `DiffBind` vignette, we have:

> By default, dba.analyze plots a correlation heatmap if it finds any
> significantly differentially bound sites, shown in Figure 3. Using
> only the differentially bound sites, we now see that the four
> tamoxifen resistant samples (representing two cell lines) cluster
> together, although the tamoxifen-responsive MCF7 replicates cluster
> closer to them than to the other tamoxifen responsive samples."

Finally, we can generate the results table, which is attached as
metadata columns to the peaks as genomic ranges. By specifying
`bCounts = TRUE`, we also obtain the normalized counts for each sample.

```{r}
tadb <- dba.report(ta2)
tadb
counts <- dba.report(ta2, bCounts=TRUE)
```

**Reproducing the log fold changes**

The following code is used only to see if we can reproduce the log fold
change obtained by the `dba.contrast` function. We extract the counts
for the top peak, and put these in the order of the samples table:

```{r}
x <- mcols(counts)[1,-c(1:6)]
x <- unlist(x)
(xord <- x[match(ta2$samples$SampleID, names(x))])
# Bioc 2.13 can use this line:
# (xord <- x[order(c(which(ta2$samples$Condition == "Resistant"),which(ta2$samples$Condition == "Responsive")))])
ta2$samples$SampleID
```

We create a vector of the conditions, and conditions combined with
tissue:

```{r}
cond <- factor(ta2$samples[,"Condition"])
condcomb <- factor(paste(ta2$samples[,"Condition"], ta2$samples[,"Tissue"]))
```

A stripchart of the counts over the conditions:

```{r}
par(mar=c(15,5,2,2))
stripchart(log(xord) ~ condcomb, method="jitter", 
           vertical=TRUE, las=2, ylab="log2 normalized counts")
```

**Finally, we show that the log2 fold change of the means is the same as
reported by the `DiffBind` functions**

```{r}
means <- tapply(xord, cond, mean)
log2(means)
log2(means[1] / means[2])
mcols(tadb)[1,]
setwd(dirwd)
```

**Footnotes** <a name="foot"></a>

**Model-based Analysis for ChIP-Seq (MACS)**

Zhang Y, Liu T, Meyer CA, Eeckhoute J, Johnson DS, Bernstein BE, Nusbaum
C, Myers RM, Brown M, Li W, Liu XS. "Model-based Analysis of ChIP-Seq
(MACS)". Genome Biol. 2008.
<http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2592715/>

[Software](http://liulab.dfci.harvard.edu/MACS/)

**Motif finding**

[Wikipedia's article on DNA sequence motifs](http://en.wikipedia.org/wiki/Sequence_motif)

A non-comprehensive list of software for motif finding:

-   [MEME/DREME](http://meme.nbcr.net/meme/)
-   [RSAT peak-motifs](http://rsat.ulb.ac.be/peak-motifs_form.cgi)
-   [motifRG(Bioconductor)](http://www.bioconductor.org/packages/release/bioc/html/motifRG.html)
-   [rGADEM(Bioconductor)](http://www.bioconductor.org/packages/release/bioc/html/rGADEM.html)

[A survey of motif finding algorithms](http://www.biomedcentral.com/1471-2105/8/S7/S21)
